stan/math/gpu/err/check_matrix_gpu.hpp

```diff
@@ -0,0 +1,222 @@
+
+    cmd_queue.enqueueNDRangeKernel(kernel_check_symmetric, cl::NullRange,
+                                   cl::NDRange(y.rows(), y.cols()),
+                                   cl::NullRange);
+
+    cmd_queue.enqueueReadBuffer(buffer_symmetric_flag, CL_TRUE, 0, sizeof(int),
+                                &symmetric_flag);
+  } catch (const cl::Error& e) {
+    check_opencl_error("symmetric_check", e);
+  }
+  //  if the matrix is not symmetric
+  if (symmetric_flag) {
+    domain_error(function, name, "is not symmetric", "");
+  }
+}
+/**
+ * Check if the specified matrix on the GPU has zeros on the diagonal
+ *
+ * @param function Function name (for error messages)
+ * @param name Variable name (for error messages)
+ * @param y Matrix to test
+ * @param pos_def_check Test if matrix y is positive definite
+ *
+ * @throw <code>std::domain_error</code> if
+ *    any diagonal element of the matrix is zero.
+ */
+inline void check_diagonal_zeros(const char* function, const char* name,
+                                 const matrix_gpu& y,
+                                 const int pos_def_check = 0) {
+  if (y.size() == 0)
+    return;
+
+  cl::Kernel kernel_check_diagonal_zeros
+      = opencl_context.get_kernel("check_diagonal_zeros");
+  cl::CommandQueue cmd_queue = opencl_context.queue();
+  cl::Context ctx = opencl_context.context();
+  int flag = 0;
+  try {
+    cl::Buffer buffer_flag(ctx, CL_MEM_READ_WRITE, sizeof(int));
+
+    cmd_queue.enqueueWriteBuffer(buffer_flag, CL_TRUE, 0, sizeof(int), &flag);
+
+    kernel_check_diagonal_zeros.setArg(0, y.buffer());
+    kernel_check_diagonal_zeros.setArg(1, y.rows());
+    kernel_check_diagonal_zeros.setArg(2, y.cols());
+    kernel_check_diagonal_zeros.setArg(3, buffer_flag);
+
+    cmd_queue.enqueueNDRangeKernel(kernel_check_diagonal_zeros, cl::NullRange,
+                                   cl::NDRange(y.rows(), y.cols()),
+                                   cl::NullRange);
+
+    cmd_queue.enqueueReadBuffer(buffer_flag, CL_TRUE, 0, sizeof(int), &flag);
+  } catch (const cl::Error& e) {
+    check_opencl_error("diag_zeros_check", e);
+  }
+  //  if zeros were found on the diagonal
+  if (flag) {
+    if (pos_def_check) {
+      domain_error(function, name, "is not positive definite", "");
+    } else {
+      domain_error(function, name, "has zeros on the diagonal.", "");
+    }
+  }
+}
+/**
+ * Check if the specified matrix on the GPU has NaN values
+ *
+ * @param function Function name (for error messages)
+ * @param name Variable name (for error messages)
+ * @param y Matrix to test
+ *
+ * @throw <code>std::domain_error</code> if
+ *    any element of the matrix is <code>NaN</code>.
+ */
+inline void check_positive_definite(const char* function, const char* name,
+                                    const matrix_gpu& y) {
+  if (y.size() == 0)
+    return;
+  check_nan(function, name, y, 1);
+  check_diagonal_zeros(function, name, y, 1);
+}
```

Comment:
Should we wait on this one and do something like what they do stan does in [`check_pos_definite()`](https://github.com/stan-dev/math/blob/c36cfa2becfcc2a0673addb1600e9133e822998d/stan/math/prim/mat/err/check_pos_definite.hpp#L33)